platform-sdk/docs/proposals/TSS-Block-Signing/tss-block-signing.md

```diff
@@ -0,0 +1,286 @@
+
+1. Public API
+2. Component Architecture
+3. Core Behaviors
+4. Configuration
+5. Metrics
+6. Performance
+
+### Public API
+
+#### Services
+
+The `TssBaseService` is extended with the following API:
+
+```Java
+/**
+ * The TssBaseService will attempt to generate TSS key material for any set candidate roster, giving it a ledger id and
+ * the ability to generate ledger signatures that can be verified by the ledger id.  Once the candidate roster has
+ * received its full TSS key material, it can be made available for adoption by the platform.
+ * </p>
+ * The TssBaseService will also attempt to generate ledger signatures by aggregating share signatures produced by
+ * calling {@link #requestLedgerSignature(byte[])}.
+ *
+ */
+public interface TssBaseService {
+
+    ...
+
+    /**
+     * Requests a ledger signature on a message hash.  The ledger signature is computed asynchronously and returned
+     * to all consumers that have been registered through {@link #registerLedgerSignatureConsumer(Consumer)}.
+     *
+     * @param messageHash The hash of the message to be signed by the ledger.
+     */
+    void requestLedgerSignature(@Nonnull byte[] messageHash);
+
+    /**
+     * Registers a consumer of pairs where the first element is the message hash and the second element is the
+     * ledger signature on the message hash.
+     *
+     * @param consumer the consumer of ledger signatures on message hashes.
+     */
+    void registerLedgerSignatureConsumer(@Nonnull final Consumer<Pair<Bytes, PairingSignature>> consumer);
+}
+```
+
+#### System Transactions
+
+The `TssShareSignatureTransaction` is a new system transaction that is only generated by the `TssBaseService` and
+not user generated. It may be possible that a node is aggregating share signature transactions from
+multiple ledger signature requests. In addition to the `share_signature`, the `roster_hash` and `share_index` are
+needed to identify the share public key to use for validation. The `message_hash` is needed to identify the message
+that was signed. A ledger signature can be produced by a threshold number of `share_signatures` for the same
+message. The threshold value for aggregation is determined by the roster indicated by the `roster_hash`.
+
+```protobuf
+message TssShareSignatureTransaction {
+  /**
+   * The hash of the roster containing the node whose share produced the share signature.
+   */
+  bytes roster_hash = 1;
+
+  /**
+   * The index of the share that produced this share signature.
+   */
+  uint64 share_index = 2;
+
+  /**
+   * The hash of the message that was signed by the share.
+   */
+  bytes message_hash = 3;
+
+  /**
+   * The share signature produced by the share.
+   */
+  bytes share_signature = 4;
+}
+```
+
+### Architecture and/or Components
+
+In addition to the extension of the `TssBaseService` interface, the service's `TransactionHandler` is extended to
+handle `TssShareSignatureTransaction` in the `pre-handle` phase.
```

Comment:
I'm curious if this can be implemented in the TssBaseService itself w/o touching any other components? Cannot a regular service pre-handle transactions w/o having to modify the TransactionHandler?

Reply:
Got it. Thanks for the clarification!